internal/clonepool/clonepool.go

```diff
@@ -0,0 +1,224 @@
+package clonepool
+
+import (
+	"fmt"
+	"github.com/labstack/gommon/log"
+	"github.com/modfin/henry/slicez"
+	"github.com/modfin/zdap"
+	"github.com/modfin/zdap/internal"
+	"github.com/modfin/zdap/internal/cloning"
+	"github.com/modfin/zdap/internal/zfs"
+	"sort"
+	"sync"
+	"time"
+)
+
+type ClonePool struct {
+	resource        internal.Resource
+	cloneContext    *cloning.CloneContext
+	ClonesAvailable int
+	claimLock       sync.Mutex
+	expireLock      sync.Mutex
+}
+
+func NewClonePool(resource internal.Resource, cloneContext *cloning.CloneContext) ClonePool {
+	return ClonePool{resource: resource, cloneContext: cloneContext}
+}
+
+func (c *ClonePool) Start() {
+	go func() {
+		for {
+			time.Sleep(time.Second)
+			c.action()
+		}
+	}()
+}
+
+func (c *ClonePool) action() {
+	c.expireLock.Lock()
+	defer c.expireLock.Unlock()
```

Comment:
I couldn't get the expire working without blocking the entire action here. Feels like there should be a way to set expiry date without blocking this much.